### PR TITLE
Handle customer stop

### LINKED ISSUE #26: Add code to handle Twilio error `21610` (customer has replied with `STOP`)
```
21610 - SMS cannot be sent to the 'To' number because the customer has replied with STOP
```

### PR DESCRIPTION
**Pull request by RDelorier**
When someone texts your twilio number with one of the messages described [here](https://www.twilio.com/help/faq/sms/does-twilio-support-stop-block-and-cancel-aka-sms-filtering) they will be added or removed from the blacklist. This pull-request intends to prevent any text from being sent to blacklisted numbers.
#### Incoming Text Messages

When someone sends one of the magic words to the app twilio will send that message through so that we can act on it. Below is the list of actions that are performed based on the body of the text message received.
- Subscribe ["START", "YES"]
  - Marks the number/friend as NOT blacklisted
- Unsubscribe ["STOP", "STOPALL", "UNSUBSCRIBE", "CANCEL", "END", "QUIT"]
  - Marks the number/friend as blacklisted

**_This required that you set up your twilio number to direct incoming text messages to {url}/receive-text**_

The receiveText action uses the TwilioRequest which verifies that the request is coming from twilio by checking the signature header against the twilio api token set in your .env file. This is disabled on the local environment so you can use your rest client to send incoming texts.
#### Outgoing Text Messages

Blacklisted numbers are excluded when sending text messages. If a text is sent that causes a BlacklistedPhoneNumberException then the number/friend will be marked as blacklisted. This can happen while sending verification texts as well as new recording notifications.
#### Displaying blacklisted numbers

Phone numbers have an appended attribute called "status" which will be one of three values
- blacklisted
- verified
- unverified

These are determined in the order above, so if a verified number is blacklisted then its status will show blacklisted and not verified.

closes #26

### CODE REVIEW

**Comment by RDelorier** [comment]
@mattstauffer I took a much more proactive approach this time. Let me know what you think!

**Comment by RDelorier** [comment]
@mattstauffer just a friendly ping in case you missed this 🔔

**Comment by mattstauffer** [comment]
@RDelorier I did!!! Geez. Sorry. Thanks :)

**Comment by mattstauffer** [comment]
OK, first reading the words of the PR.

"Subscribe ["START", "YES"]
Marks the number/friend ad blacklisted
Unsubscribe ["STOP", "STOPALL", "UNSUBSCRIBE", "CANCEL", "END", "QUIT"]
Marks the number/friend as not blacklisted"

I assume this is the opposite of what it actually does, right? Unsubscribe calls mark it as blacklisted, subscribe mark it as not blacklisted?

"This required that you set up your twilio number to direct incoming text messages to {url}/receive-text"

OK, i'm a bit nervous about this, but I'll take a look :) If you've done it in your own test panel, would you be up for showing me a walkthrough? I basically don't want to break my account while the app is live.

One last question functionality-wise: Do we notify the user if one of their people blacklists? It might be worth sending them an email so they're not expecting this to be sent to their best friend and it turns out their best friend blacklisted it.

OK, off to read the code now.

**Comment by RDelorier** [comment]
I updated the wording so it's not backwords now. I don't think you have any text handling setup on the twilio number right now so it shouldn't mess anything up since its a separate setting, I'd be happy to go over it with you, here is a screenshot of my testing setup:
screen shot 2016-07-08 at 9 27 58 am

There are no notifications setup for this feature right, we can fire an event from phoneNumber->markBlacklisted() and hook up some listeners pretty easily tho.

File: app/Http/Controllers/IncomingTextController.php [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+namespace App\Http\Controllers;
+
+use App\Http\Requests\TwilioRequest as Request;
+
+class IncomingTextController extends Controller
+{
+    protected $commands = [
+        "unsubscribe" => [
+            "STOP",
+            "STOPAll",
```
Comment (mattstauffer) [inline app/Http/Controllers/IncomingTextController.php]: Is this intentionally mixed case?

File: app/Http/Controllers/IncomingTextController.php [hunk 2]
```diff
@@ -0,0 +1,77 @@
+<?php
+
+namespace App\Http\Controllers;
+
+use App\Http\Requests\TwilioRequest as Request;
+
+class IncomingTextController extends Controller
+{
+    protected $commands = [
+        "unsubscribe" => [
```
Comment (mattstauffer) [inline app/Http/Controllers/IncomingTextController.php]: This is curious--is this a list from Twilio or are we just making up the concept of blacklisting internally? I figured listening for this happening would be listening for a webhook from Twilio?

File: app/Http/Controllers/IncomingTextController.php [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        if ($command = $this->translateCommand($request->Body)) {
+            return $this->{$command}($request);
+        }
+
+        return $this->responseMessage('Unrecognized command: '.$request->Body);
+    }
+
+    protected function subscribe(Request $request)
+    {
+        $request->phoneNumber()->markBlacklisted(false);
```
Comment (mattstauffer) [inline app/Http/Controllers/IncomingTextController.php]: Any reason we couldnt' instead use `addToBlackList` and `removeFromBlacklist`? I know it's not quite as concise, but it means we don't have to worry about boolean parameters.

File: app/Http/Controllers/IncomingTextController.php [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+            "START",
+            "YES",
+        ]
+    ];
+
+    /**
+     * Starting point for all incoming text messages.
+     */
+    public function receiveText(Request $request)
+    {
```
Comment (mattstauffer) [inline app/Http/Controllers/IncomingTextController.php:29]: What if we did this?

``` php
if ($this->commandExists($request)) {
    return $this->commandForRequest($request);
}

return $this->responseMessage('...');
```

Then those methods can be responsible for taking this syntax below and making it function in the above, little bit more readable syntax

File: app/Http/Controllers/IncomingTextController.php [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+        $request->phoneNumber()->markBlacklisted(false);
+    }
+
+    protected function unsubscribe(Request $request)
+    {
+        $request->phoneNumber()->markBlacklisted();
+    }
+
+    /**
+     * Transalate an incoming text to a command using $this->commands.
```
Comment (mattstauffer) [inline app/Http/Controllers/IncomingTextController.php]: *Translate

File: app/Http/Controllers/IncomingTextController.php [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+     *
+     * @param  string $body The body of the incoming message
+     * @return string|false
+     */
+    protected function translateCommand($body)
+    {
```
Comment (mattstauffer) [inline app/Http/Controllers/IncomingTextController.php:47]: Wonder if this could be:

``` php
protected function translateCommand($body)
{
    return collect($this->commands)->filter(function ($aliases) use ($body) {
        return in_array(strtoupper($body), $aliases);
    })->map(function ($value, $key) {
        return $key;
    })->first();
}
```

I wonder if there's a bit cleaner way than what I showed there, but I haven't had a chance to actually run it. It doesn' treturn `false`, might need restructuring, gotta run to a meeting

File: app/Http/Controllers/IncomingTextController.php (hunk 4 above)
Comment (RDelorier) [inline app/Http/Controllers/IncomingTextController.php:29]: If we also make commandForRequest default to 'unrecognized' then we can even get rid of the if and end up with

``` php
return $this->commandForRequest($request)
```

File: app/Http/Controllers/TwilioController.php [hunk 7]
```diff
@@ -2,27 +2,20 @@
 
 namespace App\Http\Controllers;
 
-use App\Http\Requests;
+use App\Http\Requests\TwilioRequest as Request;
```
Comment (mattstauffer) [inline app/Http/Controllers/TwilioController.php]: I wonder if it'd be better to instead typehint them as `TwilioRequest` and avoid the alias, so it's clearer where these magical methods are coming from to future devs?

File: app/Http/Requests/TwilioRequest.php [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+use Illuminate\Foundation\Validation\ValidatesRequests;
+
+class TwilioRequest extends Request
+{
+    /**
+     * Validate that the incoming request is from twilio.
+     *
+     * @return bool
+     */
+    public function authorize(\Services_Twilio_RequestValidator $validator)
```
Comment (mattstauffer) [inline app/Http/Requests/TwilioRequest.php]: Let's `use` this class

File: app/Http/Requests/TwilioRequest.php [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+    {
+        if (app()->isLocal()) {
+            return true;
+        }
+
+        return $validator->validate(
```
Comment (mattstauffer) [inline app/Http/Requests/TwilioRequest.php]: Nice work here, I literally didn't even know this existed

File: app/Http/Requests/TwilioRequest.php [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+    }
+
+    public function rules()
+    {
+        return [];
+    }
+
+    public function phoneNumber()
+    {
+        $number = PhoneNumber::byNumber($this->input('From'))->first();
```
Comment (mattstauffer) [inline app/Http/Requests/TwilioRequest.php]: This was previously using "findVerifiedByTwilioNumber".. now it's just using `byNumber`. Are we going to miss them because they need to be translated, or has there been a change to the syntax?

File: app/Http/Requests/TwilioRequest.php [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+
+        return $number ?: Friend::byNumber($this->input('From'))->first();
+    }
+
+    public function isFromVerifiedNumber()
+    {
+        if ($number = $this->phoneNumber()) {
```
Comment (mattstauffer) [inline app/Http/Requests/TwilioRequest.php]: could we do this?

``` php
return $this->phoneNumber() && $this->phoneNumber()->is_verified;
```

File: app/Jobs/NotifyFriendsOfRecording.php [hunk 12]
```diff
@@ -5,29 +5,36 @@
 use App\Jobs\Job;
 use App\Phone\TwilioClient;
 use App\PhoneNumber;
+use App\Recording;
 use Illuminate\Log\Writer as Logger;
+use Illuminate\Support\Fluent;
+use App\Phone\Exceptions\BlacklistedPhoneNumberException;
```
Comment (mattstauffer) [inline app/Jobs/NotifyFriendsOfRecording.php]: Alphabetize `use` please :D

File: app/PhoneNumber.php [hunk 13]
```diff
@@ -27,27 +30,40 @@ public function markVerified()
...
+        info('Number was blacklisted', ['class' => get_called_class(), 'id' => $this->id]);
     }
 
     public function scopeVerified(Builder $builder)
     {
         return $builder->where('is_verified', true);
     }
 
-    public static function findByTwilioNumber($number)
+    public function scopeBlacklisted(Builder $builder, $value = true)
```
Comment (mattstauffer) [inline app/PhoneNumber.php:56]: Are there any places in the site where we need to be default scoping to "not blacklisted" that we're not? Or do you have that all covered?

**Comment by mattstauffer** [comment]
This is fantastic work, dude. Thank you for all of this!

File: app/PhoneNumber.php (hunk 13 above)
Comment (RDelorier) [inline app/PhoneNumber.php:56]: I believe I added it where needed but I'll double check tonight

**Comment by mattstauffer** [comment]
@RDelorier I know I'm being dense, but I onl yhave a tiny bit of time and brain power left for this today: Where is that panel you screenshotted? Did you go to the messaging section and then Create a new Messaging Service?

Thanks!

File: app/Http/Controllers/TwilioController.php (hunk 7 above)
Comment (RDelorier) [inline app/Http/Controllers/TwilioController.php]: Good point, I even confused myself just now with this.

File: app/Http/Requests/TwilioRequest.php (hunk 9 above)
Comment (RDelorier) [inline app/Http/Requests/TwilioRequest.php]: Source diving for the win 🎉

File: app/Http/Requests/TwilioRequest.php (hunk 10 above)
Comment (RDelorier) [inline app/Http/Requests/TwilioRequest.php]: The byNumber scope does a str_replace on all values passed in since we never store the +1, now it will just work with or without the prefix.

File: app/Http/Controllers/IncomingTextController.php (hunk 2 above)
Comment (RDelorier) [inline app/Http/Controllers/IncomingTextController.php]: Twilio handles the blacklisting automatically and then they send the message through so that we can act on it, I didn't see any specific webhook setup but the link in the pr explains how it works.

File: app/Http/Controllers/IncomingTextController.php (hunk 3 above)
Comment (RDelorier) [inline app/Http/Controllers/IncomingTextController.php]: I just did it this way so there would be a single point where we could dispatch events, I'll add those methods as aliases to markBlacklisted.

**Comment by RDelorier** [comment]
@mattstauffer I just went through the [Phone Numbers](https://www.twilio.com/console/phone-numbers/dashboard) on the dashboard and clicked the phone number

**Comment by mattstauffer** [comment]
Got it. Thanks!

**Comment by RDelorier** [comment]
Should tests/\* be in the nitpick.json? It makes kinda hard to follow these pull requests with all those snake_case methods.

**Comment by mattstauffer** [comment]
Yah I think thats the only option for now. I hate losing nitpick on the
tests but otherwise it's unusable.

Any plans to ignore specific rules yet @adamwathan? :)

**Comment by adamwathan** [comment]
Nah not any time soon.

**Comment by mattstauffer** [comment]
@RDelorier just dropped tests dir from nitpick for now; you can merge master if you want.

https://github.com/mattstauffer/pulledover/pull/33

**Comment by RDelorier** [comment]
zzz of course I messed it up lol

**Comment by RDelorier** [comment]
@mattstauffer yay no more nitpick onslaught :)

File: app/Listeners/NotifyOwnerFriendWasBlacklisted.php [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+
+    /**
+     * Handle the event.
+     *
+     * @param  FriendWasBlacklisted  $event
+     * @return void
+     */
+    public function handle(FriendWasBlacklisted $event)
+    {
+        // todo send owner notification
```
Comment (RDelorier) [inline app/Listeners/NotifyOwnerFriendWasBlacklisted.php:29]: @mattstauffer Do you want to send out an email or a text message here?

**Comment by mattstauffer** [comment]
@RDelorier OK, back in the office after the baby showing up. Hope to get to this very soon--thanks as always for patience!

File: app/Http/Controllers/TwilioController.php [hunk 15]
```diff
@@ -47,20 +40,18 @@ private function promptToRegister($request)
...
 
         return $this->hangup();
     }
 
     private function saveRecording($request)
     {
-        $number = PhoneNumber::findByTwilioNumber($request->input('Caller'));
-
-        $recording = new Recording([
+        return $request->phoneNumber()->user->recordings()->create([
```
Comment (mattstauffer) [inline app/Http/Controllers/TwilioController.php:54]: 👌  clean

File: app/Http/Controllers/TwilioController.php [hunk 16]
```diff
@@ -2,27 +2,20 @@
...
+    public function callHook(TwilioRequest $request)
     {
-        try {
-            $phoneNumber = PhoneNumber::findVerifiedByTwilioNumber($request->input("From"));
-        } catch (Exception $e) {
-            return $this->promptToRegister($request);
-        }
-
-        return $this->startRecording();
+        return $request->isFromVerifiedNumber() ? $this->startRecording() : $this->promptToRegister();
```
Comment (mattstauffer) [inline app/Http/Controllers/TwilioController.php:18]: :+1:

File: app/Http/Requests/TwilioRequest.php [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+    public function rules()
+    {
+        return [];
+    }
+
+    public function phoneNumber()
+    {
+        if (is_null($this->phoneNumber)) {
+            //set the number to first [PhoneNumber | Friend | NULL]
+            $this->phoneNumber = PhoneNumber::byNumber($this->input('From'))->first()
```
Comment (mattstauffer) [inline app/Http/Requests/TwilioRequest.php:41]: Will we have any collisions if someone both has an account and also is someone else's friend?

File: app/Listeners/NotifyOwnerFriendWasBlacklisted.php (hunk 14 above)
Comment (mattstauffer) [reply app/Listeners/NotifyOwnerFriendWasBlacklisted.php:29]: Let's do email 👌

File: app/Http/Requests/TwilioRequest.php (hunk 17 above)
Comment (RDelorier) [reply app/Http/Requests/TwilioRequest.php:41]: I don't think it will since `PhoneNumber` is checked before `Friend` but I'll have to add a test to be sure.

That does bring up another issue, it doesn't look like I added a way for the controller to restrict to `PhoneNumber` when accepting calls.

**Comment by mattstauffer** [comment]
I wasn't 100% able to follow everything but what i could follow looks good and the test look god. So .. holler when you feel like that PhoneNumber thing is handled and then let's do this. Thanks for all your work!!

**Comment by RDelorier** [comment]
Your welcome and thanks for taking time out to review everything. I'm going to leave a checklist here so I don't forget. I'll ping you when it's ready to roll out.
- [ ] [Send email on blacklisted](https://github.com/mattstauffer/pulledover/pull/31#discussion_r76896107)
- [ ] [Add a way to restrict the phone number to model](https://github.com/mattstauffer/pulledover/pull/31#discussion_r76895897)

**Comment by mattstauffer** [comment]
@RDelorier Hey dude--I know I took *forever* on this. I want to follow up and see if we can get this thing finished up. Do you think it;s something you can find time for? If not, I'll be happy to try to take over and get it done. Just let me know. Thanks!

**Comment by RDelorier** [comment]
Hey Matt, I've been pretty swamped with work lately and have a few more busy weeks ahead so I don't think I'll have time to finish this up for a while. If you have any questions feel free to ping me and I'll take a look to see if it jogs my memory.

**Comment by RDelorier** [comment]
I think it's safe to close this, I don't even remember what changes were made our why. Probably just discouraging other contributions by leaving it here.
